### PR TITLE
Add htseqclip tool

### PR DESCRIPTION
**Pull request by michauhl**
Add Galaxy wrapper for htseq-clip, a toolset for the analysis of eCLIP/iCLIP datasets:
https://github.com/EMBL-Hentze-group/htseq-clip

### CODE REVIEW

**Comment by michauhl** [comment]
Tests are fine locally, but here it seems pysam.index() inside python script cannot create a BAM index file for the input test-data BAM files (probably write-protected ?):

pysam.utils.SamtoolsError: 'samtools returned with error 1: stdout=, stderr=samtools index: failed to create or write index\n'

How to solve this? Normally input BAM files come without index files, but index files need to created before using htseq-clip, otherwise it doesn't work. Or is this not needed inside Galaxy / are they created automatically ?

**Comment by michauhl** [comment]
Oh well, the tool seems to be sensitive to computing environment (maybe tool versions), resulting in different row ordering in the output table files (therefore test file equals tests fail). Need to ask developer why this is the case ...

File: tools/rna_tools/htseq-clip/htseq_clip.xml [hunk 1]
```diff
@@ -35,7 +35,7 @@
                 --hce-g $action_type.hce_options.hce_g
                 --hce-q $action_type.hce_options.hce_q
                 $action_type.hce_options.hce_primary
-                --hce-c 4
+                --hce-c 1
```
Comment (bgruening) [inline tools/rna_tools/htseq-clip/htseq_clip.xml:38]: this should use GALAXY_SLOTS I think you want to make it configurable, correct?

**Comment by bgruening** [comment]
> Oh well, the tool seems to be sensitive to computing environment (maybe tool versions), resulting in different row ordering in the output table files (therefore test file equals tests fail). Need to ask developer why this is the case ...

You can add sort="true" to test-outputs https://docs.galaxyproject.org/en/latest/dev/schema.html#id106

File: tools/rna_tools/htseq-clip/htseq_clip.xml (hunk 1 above)
Comment (michauhl) [reply tools/rna_tools/htseq-clip/htseq_clip.xml:38]: Yes, I was going to ask how to adjust this for Galaxy ...  :)

Comment (michauhl) [reply tools/rna_tools/htseq-clip/htseq_clip.xml:38]: How to write this here ? I looked in some wrappers and it is always written like:
` --cpus "\${GALAXY_SLOTS:-12}"`
What does the number mean ? Max CPUs supported ?

Comment (bgruening) [reply tools/rna_tools/htseq-clip/htseq_clip.xml:38]: That is bash syntax, it's means 12 if the variable is not set. A default vakue

Comment (michauhl) [reply tools/rna_tools/htseq-clip/htseq_clip.xml]: Then I'll give it a go with `--hce-c \${GALAXY_SLOTS:-1}`

File: tools/rna_tools/htseq-clip/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  htseq-clip is a toolset for the analysis of eCLIP/iCLIP datasets
+homepage_url: https://github.com/EMBL-Hentze-group/htseq-clip
+long_description:
+  htseq-clip is a toolset for the analysis of eCLIP/iCLIP datasets. 
+  It can be used to generate files necessary for data analysis using 
+  the companion R/Bioconductor package DEWSeq.
+remote_repository_url: https://github.com/bgruening/galaxytools/tree/master/tools/rna_tools/htseq-clip
+type: unrestricted
+categories:
+- Sequence Analysis
```
Comment (bgruening) [inline tools/rna_tools/htseq-clip/.shed.yml:13]: Those needs to be toolshed categories.... You can not add arbritrairy ones.

File: tools/rna_tools/htseq-clip/htseq_clip.xml [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+                </section>
+
+            </when>
+
+            <when value="create_count_table">
+                <repeat name="exp_bams" min="1" title="CLIP-seq experiment BAM inputs">
+                    <param name="exp_bam" type="data" format="bam" label="BAM files belonging to the CLIP-seq experiment" help="Select BAM file belonging to the CLIP-seq experiment. NOTE that order determines replicate numbering in output tables"/>
+                </repeat>
+                <repeat name="ctr_bams" min="1" title="CLIP-seq control BAM inputs">
+                    <param name="ctr_bam" type="data" format="bam" label="BAM files belonging to the CLIP-seq control" help="Select BAM file belonging to the CLIP-seq control. NOTE that order determines replicate numbering in output tables"/>
```
Comment (bgruening) [inline tools/rna_tools/htseq-clip/htseq_clip.xml:94]: You can simply use multiple=true and avoid the repeat

File: tools/rna_tools/htseq-clip/.shed.yml (hunk 2 above)
Comment (michauhl) [reply tools/rna_tools/htseq-clip/.shed.yml:13]: There are no categories for CLIP-seq or RNA-binding proteins. Then I will just keep RNA and Sequence analysis.

File: tools/rna_tools/htseq-clip/htseq_clip.xml (hunk 3 above)
Comment (michauhl) [reply tools/rna_tools/htseq-clip/htseq_clip.xml:94]: Wolfgang suggested this in the tech tips chat, as I want to keep the order of the input files (rep1, rep2 ... ).

**Comment by michauhl** [comment]
@bgruening  From my side it can now be merged. Regarding RAM usage, the developer suggested >= 32GB, 64GB would be ideal. Regarding toolshed categories, is there a way to add categories (CLIP-seq?) for CLIP-seq related tools? Since there are a number of tools available on Galaxy by now.

**Comment by bgruening** [comment]
> is there a way to add categories (CLIP-seq?) for CLIP-seq related tools? Since there are a number of tools available on Galaxy by now.

I think we can do that.

@martenson can you help here?

**Comment by martenson** [comment]
@bgruening @michauhl Yeah, we can do that. We usually create one on TTS first and once there are actual tools present we create the MTS one. I've created the category on TTS(`CLIP-seq`) already, so please ping me when you have tools there ready for MTS. You can also supply description of such category, I've chosen something generic for now.

**Comment by bgruening** [comment]
If we merge this is gets pushed to TTS and TS. So I guess it makes sense to have one on TS as well?

File: tools/rna_tools/htseq-clip/.shed.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+description: |
+  htseq-clip is a toolset for the analysis of eCLIP/iCLIP datasets
+homepage_url: https://github.com/EMBL-Hentze-group/htseq-clip
+long_description:
+  htseq-clip is a toolset for the analysis of eCLIP/iCLIP datasets. 
+  It can be used to generate files necessary for data analysis using 
+  the companion R/Bioconductor package DEWSeq.
+remote_repository_url: https://github.com/bgruening/galaxytools/tree/master/tools/rna_tools/htseq-clip
+type: unrestricted
+categories:
```
Comment (martenson) [inline tools/rna_tools/htseq-clip/.shed.yml:12]: @bgruening do you mean that you want to put `CLIP-seq` here already?

Comment (bgruening) [reply tools/rna_tools/htseq-clip/.shed.yml:12]: @michauhl this is what I understood, correct?

Comment (michauhl) [reply tools/rna_tools/htseq-clip/.shed.yml:12]: @bgruening @martenson if possible, yes.

Comment (martenson) [reply tools/rna_tools/htseq-clip/.shed.yml:12]: Sounds great, done.

Comment (bgruening) [reply tools/rna_tools/htseq-clip/.shed.yml:12]: @michauhl check out https://toolshed.g2.bx.psu.edu/ and add your preferred categories :)

Comment (michauhl) [reply tools/rna_tools/htseq-clip/.shed.yml:12]: @martenson @bgruening perfect, thank you! :)

**Comment by michauhl** [comment]
@bgruening can you merge the pull request? From my side everything looks fine now.

**Pull request merged**